addons/mrp/models/mrp_production.py

```diff
@@ -624,54 +623,72 @@ def button_plan(self):
             quantity = order.product_uom_id._compute_quantity(order.product_qty, order.bom_id.product_uom_id) / order.bom_id.product_qty
             boms, lines = order.bom_id.explode(order.product_id, quantity, picking_type=order.bom_id.picking_type_id)
             order._generate_workorders(boms)
-        self.plan_workorders()
+            order._plan_workorders()
         return True
 
     def _get_start_date(self):
-        return self.date_start_wo or datetime.now()
+        return self.date_start_wo or datetime.datetime.now()
 
-    def plan_workorders(self):
-        WorkOrder = self.env['mrp.workorder']
-        ProductUom = self.env['uom.uom']
-        for order in self.filtered(lambda x: x.state == 'planned'):
-            order.workorder_ids.write({'date_planned_start': False, 'date_planned_finished': False})
+    def _plan_workorders(self):
+        """ Plan all the production's workorders depending on the workcenters
+        work schedule"""
+        self.ensure_one()
 
         # Schedule all work orders (new ones and those already created)
-        for order in self:
-            start_date = order._get_start_date()
-            from_date_set = False
-            for workorder in order.workorder_ids:
-                workcenter = workorder.workcenter_id
-                wos = WorkOrder.search([('workcenter_id', '=', workcenter.id), ('date_planned_finished', '<>', False),
-                                        ('state', 'in', ('ready', 'pending', 'progress')),
-                                        ('date_planned_finished', '>=', start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT))], order='date_planned_start')
-                from_date = start_date
-                to_date = workcenter.resource_calendar_id.attendance_ids and workcenter.resource_calendar_id.plan_hours(workorder.duration_expected / 60.0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                if to_date:
-                    if not from_date_set:
-                        # planning 0 hours gives the start of the next attendance
-                        from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                        from_date_set = True
+        start_date = self._get_start_date()
+        for workorder in self.workorder_ids:
```

Comment:
you shouldn't rely on this! self.workorder_ids may not return the workorders correctly ordered
either you sort them by next_workorder_id before looping over or you enforce an order on the mrp.workorder model (currently none is enforced)

Reply:
Seems that the default order is good enough: 
https://github.com/odoo/odoo/blob/086b9ad879664448d8a4c9102b2cf158def04444/odoo/models.py#L249